# Monte Carlo Ising Model

- Magnetism is the sum of the spins
- averages of an observable: 1/Q * sum of obsevable * probabiity
- 2x2 lattice, 16 possible configs: 2^N
- only calaculte states with considerable boltzmann factors: importance sampling, rather than sampling states randomly - how MC method works
- not a time based method, just about minimisng energy: a MC step is not a time step its a "trial step" 
- MC uses canonical ensemble
- minimise Helmhotlz free energy

### Task 1

Task 1a:  
no. nearest neighbours = 2D where D is the dimension  
min energy when all spins are aligned  
2 configs where the spins are all aligned (all down or all up) therefore multiplicity = 2
Entropy: 
s = kbln(W) = kbln(2)

TASK 1b:  
D = 3, N = 1000  
1000 positions and 2 spins states therefore 2000 configs or  
W = 2 * N!/n!(N-n)! = 2000  
change in S: kbln(2000) - kbln(2) = kbln(1000)  
  
E of one interacting spin = -6J  
E of one interacting oppsite spin = 6J  
change in energy = 12J  
factor of 1/2 removed as no double counting for 1 molecule


Task 1c:  
low T - energetic driving force dominates and most spins parallel  
high T - entropic driving force dominates and spins rotate freely  
1D: M = 3 - 2 = 1  
2D: M = 13 - 12 = 1  
absolute zero: all spins aligned  
M = +- 1000

### Task 2

In [1]:
import numpy as np

Task 2a:

In [597]:
%load_ext autoreload
%autoreload 2

The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload


In [601]:
import IsingLattice as il
%run ILcheck.py

### Task 3

Task 3a:  
2 types of spin, 100 spins: 2^100 configs  
time = 2^100/1x10^9 = 1.3x10^21 s

Task 3b:  


In [4]:
%matplotlib

Using matplotlib backend: MacOSX


In [602]:
%matplotlib
%run ILanim.py


Using matplotlib backend: MacOSX


UnboundLocalError: local variable 'avg_E' referenced before assignment

Task 3c:  
Below Tc minimisation of energy dominates and the majority of spins are aligned, therefore <M> != 0

Explain why average it tends to is wrong for average energy and M  
takes average value from beginning to end, not from when it equilibrates

<div>
<img src="Task%203c.png" width="400"/>
</div>

In [14]:
# at low T all the spins should be aligned at equilibrium
# sometimes stuck in a metastable state (local minima) as on interface between
# opposite spins there are unfavourable interactions therefore cant be 
# global minima
# dependent on startin position

### Task 4

Task 4a:  
recording 5x time to do 2000 steps

In [5]:
time_list = []

In [78]:
%run ILtimetrial.py

2000 steps took 41.23636899999997 s


In [6]:
for i in range(0,5):
    %run ILtimetrial.py
    time_list.append(elapsed_time)


2000 steps took 41.75272 s
2000 steps took 41.916987 s
2000 steps took 41.88158399999999 s
2000 steps took 41.60669300000001 s
2000 steps took 41.362611000000015 s


In [13]:
avg_time = np.average(time_list)
error = np.std(time_list)/np.sqrt(len(time_list))
print(avg_time)
print(error)

41.704119000000006
0.09061748360884529


41.70 +- 0.09s

Task 4b:  
now have 0 for loops instead of 2, only call energy 2 times not 4

np.roll(arr, n) shifts the array by n to the right, wrapping round the edges and putting the last element to the start.  
for the lattice we can roll the whole lattice to the right and also up. this means we can remove both for loops and removes the need to loop over coordinates.

Task 4c:

In [21]:
%run IsingLattice.py
%run ILtimetrial.py

2000 steps took 0.169354000000002 s


In [10]:
time_list_fast = []
for i in range(0,5):
    %run ILtimetrial.py
    time_list_fast.append(elapsed_time)

2000 steps took 0.16250700000000506 s
2000 steps took 0.15863099999998553 s
2000 steps took 0.16045400000001564 s
2000 steps took 0.15831399999999007 s
2000 steps took 0.15939000000000192 s


In [11]:
avg_time_fast = np.average(time_list_fast)
error_fast = np.std(time_list_fast)/np.sqrt(len(time_list_fast))
print(avg_time_fast)
print(error_fast)

0.15985919999999965
0.0006776695005711805


0.1721 +- 0.0020s

In [603]:
lattice = np.random.choice([-1,1], size=(5, 5))

In [604]:
lattice

array([[ 1, -1,  1,  1, -1],
       [-1,  1,  1,  1,  1],
       [-1, -1,  1, -1, -1],
       [-1,  1,  1, -1,  1],
       [-1,  1, -1,  1,  1]])

In [606]:
R = np.roll(lattice, shift=-1, axis=0)
UP = np.roll(lattice, shift=-1, axis=1)

In [609]:
R

array([[-1,  1,  1,  1,  1],
       [-1, -1,  1, -1, -1],
       [-1,  1,  1, -1,  1],
       [-1,  1, -1,  1,  1],
       [ 1, -1,  1,  1, -1]])

In [608]:
UP

array([[-1,  1,  1, -1,  1],
       [ 1,  1,  1,  1, -1],
       [-1,  1, -1, -1, -1],
       [ 1,  1, -1,  1, -1],
       [ 1, -1,  1,  1, -1]])

In [611]:
lattice * (R + UP)

array([[-2, -2,  2,  0, -2],
       [ 0,  0,  2,  0, -2],
       [ 2, -2,  0,  2,  0],
       [ 0,  2, -2, -2,  0],
       [-2, -2, -2,  2, -2]])

In [613]:
sum_sisj = np.sum(lattice * (R + UP), axis=(0, 1))
sum_sisj

-10

### Task 5a

In [ ]:
# used curve to determine what period we ignore
# anything under the curve is definitely at equilibrium
# put grid size being used into our funciton to find when it reached equilibrium.

In [131]:
%run ILfinalframe.py

Step  0
Step  100
Step  200
Step  300
Step  400
Step  500
Step  600
Step  700
Step  800
Step  900
Step  1000
Step  1100
Step  1200
Step  1300
Step  1400
Step  1500
Step  1600
Step  1700
Step  1800
Step  1900
Step  2000
Step  2100
Step  2200
Step  2300
Step  2400
Step  2500
Step  2600
Step  2700
Step  2800
Step  2900
Step  3000
Step  3100
Step  3200
Step  3300
Step  3400
Step  3500
Step  3600
Step  3700
Step  3800
Step  3900
Step  4000
Step  4100
Step  4200
Step  4300
Step  4400
Step  4500
Step  4600
Step  4700
Step  4800
Step  4900
Step  5000
Step  5100
Step  5200
Step  5300
Step  5400
Step  5500
Step  5600
Step  5700
Step  5800
Step  5900
Step  6000
Step  6100
Step  6200
Step  6300
Step  6400
Step  6500
Step  6600
Step  6700
Step  6800
Step  6900
Step  7000
Step  7100
Step  7200
Step  7300
Step  7400
Step  7500
Step  7600
Step  7700
Step  7800
Step  7900
Step  8000
Step  8100
Step  8200
Step  8300
Step  8400
Step  8500
Step  8600
Step  8700
Step  8800
Step  8900
Step  9000
Step  9100


In [13]:
# 8x8 T=1

<div>
<img src="task%205a%208x8%20T=1.png" width="400"/>
</div>

In [5]:
import matplotlib.pyplot as plt
from scipy.optimize import curve_fit

<div>
<img src="task%205a%20meta%20stable%20state.png" width="400"/>
</div>

In [37]:
# 32x32 in a meta stable state 
# interface formed - increases entropy
# will be very hard to get out this local minima at this temperature (1K)
# as will need to get rid of interface and flip back many spins which is 
# entropically unfavourable. will need to calculate the free energy to 
# properly understand.
# also interface is not curved, therefore not as stable as can be

In [646]:
grid = np.array([8,10,12,14,16,18,20, 25, 32])
steps = np.array([600, 950, 1500, 4600, 10000, 14000, 25000, 50000, 150000])
steps2 = np.array([300,900,2800,4200,10000,14400,23000, 45000, 160000])
steps3 = np.array([786, 1470, 2000,4200,10600,22000,26000, 40000, 120000])
steps_adj = steps/grid**2
steps2_adj = steps2/grid**2
steps3_adj = steps3/grid**2
stepsavg = (steps+steps2+steps3)/3
steps_grid_adjusted = stepsavg/grid**2
error_steps = []
error_steps_grid_adjusted = []

for i in range(len(steps)):
    arr = np.array([steps[i], steps2[i], steps3[i]])
    error_steps.append(np.std(arr)/np.sqrt(3))
    
for i in range(len(steps_grid_adjusted)):
    arr = np.array([steps_adj[i], steps2_adj[i], steps3_adj[i]])
    error_steps_grid_adjusted.append(np.std(arr)/np.sqrt(3))
    

<font color='red'>investigate different temperatures</font>

In [777]:
# def f(x, a, b):
#     return a + b*x**2
# def f(x, a, b, c):
#     return a + b*np.exp(c*x)
def f(x, a, b, n):
    return a + b*x**n

def g(x):
    return 10**(-0.52)*x**4.11

popt, pcov = curve_fit(f, grid, stepsavg)

plt.errorbar(grid, stepsavg, yerr=error_steps, fmt='o', linestyle='-')
#plt.plot(grid, stepsavg, ls = '', marker ='^')
plt.plot(np.linspace(8,32,1000), f(np.linspace(8,32,1000), *popt)+15000, label = 'y = {:.2f} + {:.2f}x^{:.2f}'.format(popt[0]+15000, popt[1], popt[2]))
plt.plot(np.linspace(8,32,1000), g(np.linspace(8,32,1000)))
plt.xlabel('Grid size')
plt.ylabel('Steps to reach equilibrium')
plt.legend(fontsize=12, loc=2)


In [679]:
10**(-0.52)

0.3019951720402016

In [165]:
error_steps_grid_adjusted

[1.806203142414127,
 1.4879765031995882,
 2.1466709480314123,
 0.5554398509712426,
 0.637887953849786,
 6.558591543419437,
 1.8002057495577393,
 3.771236166328254,
 9.583073856692543]

In [647]:
error_log = []
for i in range(len(steps_grid_adjusted)):
    arr = np.array([steps_adj[i], steps2_adj[i], steps3_adj[i]])
    error_log.append(np.std(arr)/np.average(arr))
    
# error propagation for logs is std/avg

In [649]:
error_log = []
for i in range(len(steps)):
    arr = np.array([steps[i], steps2[i], steps3[i]])
    error_log.append(np.std(arr)/np.average(arr))

In [773]:
def f(x,m,c):
    return m*x + c

popt, pcov = curve_fit(f, np.log10(grid), np.log10(stepsavg))

plt.errorbar(np.log10(grid), np.log10(stepsavg), yerr=error_log, fmt='o', linestyle='')
#plt.plot(grid, steps_grid_adjusted, ls = '', marker ='^')
plt.plot(np.linspace(0.9,1.5,1000), f(np.linspace(0.9,1.5,1000), popt[0], popt[1]+0.5), label = 'y={:.2f}x + {:.2f}'.format(popt[0], popt[1]+0.5))
#plt.plot(np.linspace(2,3.5,1000), f(np.linspace(2,3.5,1000), popt[0], popt[1]+0.45), label = 'y={:.2f}x + {:.2f}'.format(popt[0], popt[1]+0.5))
plt.xlabel('log Grid size')
plt.ylabel('log Steps to reach equilibrium')
plt.legend(fontsize=15)


In [774]:
# define an equation to calculate number of steps to reach equilibrium as a function of grid size
def steps_to_equilibrium(grid_size, T):
    return 10**(popt[1]+0.5)*(grid_size**popt[0])
steps_to_equilibrium(8,1)

1558.6081347211295

<font color='red'> need to do 10^(popt[0]*x + popt[1]+0.45) when adding</font>

In [767]:
grid_mutli = np.array([4,8,16,32])
T = np.array([0.5,1,1.5])

steps = np.array([[50, 400, 6000, 175000],
                 [50,600, 10000, 150000], 
                 [54,800, 12000, 170000]])
steps2 = np.array([[65, 600, 9000, 150000],
                  [43, 300,10000, 160000],
                  [42,1000, 14000, 150000]])
steps3 = np.array([[47, 300, 8500, 75000],
                  [40, 786,10600, 120000],
                  [55,1800, 13000, 200000]])

In [768]:
steps_0_5_avg = (steps[0] + steps2[0] + steps2[0])/(3*grid_mutli)
steps_1_avg = (steps[1] + steps2[1] + steps2[1])/(3*grid_mutli)
steps_1_5_avg = (steps[2] + steps2[2] + steps2[2])/(3*grid_mutli)

steps_0_5_error = []
steps_1_error = []
steps_1_5_error = []

#logged error propagation

for i in range(len(steps[0])):
    arr = np.array([steps[0][i], steps2[0][i], steps3[0][i]])
    steps_0_5_error.append(np.std(arr)/np.average(arr))
for i in range(len(steps[1])):
    arr = np.array([steps[1][i], steps2[1][i], steps3[1][i]])
    steps_1_error.append(np.std(arr)/np.average(arr))
for i in range(len(steps[2])):
    arr = np.array([steps[2][i], steps2[2][i], steps3[2][i]])
    steps_1_5_error.append(np.std(arr)/np.average(arr))
    


In [769]:
steps_0_5_error

[0.14581496062984836,
 0.2878197989826109,
 0.1675320824257832,
 0.31868871959954903]

In [772]:
def f(x,m,c):
    return m*x + c

popt0, pcov0 = curve_fit(f, np.log10(grid_mutli), np.log10(steps_0_5_avg))
popt1, pcov1 = curve_fit(f, np.log10(grid_mutli), np.log10(steps_1_avg))
popt2, pcov2 = curve_fit(f, np.log10(grid_mutli), np.log10(steps_1_5_avg))

plt.errorbar(np.log10(grid_mutli), np.log10(steps_0_5_avg), yerr=steps_0_5_error, fmt='o', linestyle='', c='k')
plt.plot(np.linspace(0.6,1.5,1000), f(np.linspace(0.6,1.5,1000), popt0[0], popt0[1]), label = 'T = 0.5 K', c='k')

plt.errorbar(np.log10(grid_mutli), np.log10(steps_1_avg), yerr=steps_1_error, fmt='o', linestyle='', c='orange')
plt.plot(np.linspace(0.6,1.5,1000), f(np.linspace(0.6,1.5,1000), popt1[0], popt1[1]), label = 'T = 1 K', c='orange')

plt.errorbar(np.log10(grid_mutli), np.log10(steps_1_5_avg), yerr=steps_1_5_error, fmt='o', linestyle='', c='green')
plt.plot(np.linspace(0.6,1.5,1000), f(np.linspace(0.6,1.5,1000), popt2[0], popt2[1]), label = 'T = 1.5 K', c='green')

plt.xlabel('log Grid size')
plt.ylabel('log Steps to reach equilibrium')
plt.legend(fontsize=15)

<div>
<img src="task%205a%20different%20T.png" width="400"/>
</div>

In [771]:
def f(x, a, b, n):
    return a + b*x**n

popt0, pcov0 = curve_fit(f, grid_mutli, steps_0_5_avg)
popt1, pcov1 = curve_fit(f, grid_mutli, steps_1_avg)
popt2, pcov2 = curve_fit(f, grid_mutli, steps_1_5_avg)

plt.errorbar(grid_mutli, steps_0_5_avg, yerr=steps_0_5_error, fmt='o', linestyle='', c='k')
plt.plot(np.linspace(2,32,1000), f(np.linspace(2,32,1000), popt0[0], popt0[1], popt0[2]), label = 'T = 0.5 K', c='k')

plt.errorbar(grid_mutli, steps_1_avg, yerr=steps_1_error, fmt='o', linestyle='', c='orange')
plt.plot(np.linspace(2,32,1000), f(np.linspace(2,32,1000), popt1[0], popt1[1], popt1[2]), label = 'T = 1 K', c='orange')

plt.errorbar(grid_mutli, steps_1_5_avg, yerr=steps_1_5_error, fmt='o', linestyle='', c='green')
plt.plot(np.linspace(2,32,1000), f(np.linspace(2,32,1000), popt2[0], popt2[1], popt2[2]), label = 'T = 1.5 K', c='green')

plt.xlabel('Grid size')
plt.ylabel('Steps to reach equilibrium')
plt.legend(fontsize=15)

In [ ]:
# number of steps to reach equilibrium, where one step is when every spin in 
# the lattice has been flipped - a Monte Carlo cycle (individual ones are trial moves)

In [168]:
# shifted up so equilibrium will definitely have been reached

In [18]:
cutoff = int(f(32, popt[0], popt[1], popt[2])) + 15000
cutoff

157287

<div>
<img src="task%205a%20quadratic.png" width="400"/>
</div>

In [ ]:
# cutoff is defined in program as a shifted version of the function used to fit
# the curve. this is to make sure the no steps vs gridsize lies completely under
# this curve
# new avg values now recorded after the cuttoff (when at equilibrium)

In [101]:
# at high temps, the average magnetism is zero therefore the system is disordered and above
# the curie T

### Task 5b

TASK 5b: Use ILtemperaturerange.py to plot the average energy and magnetisation for each temperature, with error bars, for an 8\times 8 lattice. Use your intuition and results from the script ILfinalframe.py to estimate how many cycles each simulation should be. The temperature range 0.25 to 5.0 is sufficient. Use as many temperature points as you feel necessary to illustrate the trend, but do not use a temperature spacing larger than 0.5. The NumPy function savetxt() stores your array of output data on disk — you will need it later. Save the file as 8x8.dat so that you know which lattice size it came from.

In [10]:
# 8x8 reaches equilibrium in under 1000 steps therefore dont need many cycles
# T range 0.5-5 0.25 incriment, 20000 runs

<div>
<img src="task%205b%208x8%200.5-5.0%200.1.png" width="400"/>
</div>

In [667]:
T = np.loadtxt("8x8.dat")[:,0]
E_spin = (np.loadtxt("8x8.dat")[:,1] + np.loadtxt("8x8_2.dat")[:,1] + np.loadtxt("8x8_3.dat")[:,1])/(3*64)
M_spin =  (np.loadtxt("8x8.dat")[:,3] + np.loadtxt("8x8_2.dat")[:,3] + np.loadtxt("8x8_3.dat")[:,3])/(3*64)

E_spin_error = []
for i in range(len(np.loadtxt("8x8.dat")[:,1])):
    arr = np.array([np.loadtxt("8x8.dat")[:,1][i], np.loadtxt("8x8_2.dat")[:,1][i], np.loadtxt("8x8_3.dat")[:,1][i]])/64
    E_spin_error.append(np.std(arr)/np.sqrt(3))
    
M_spin_error = []
for i in range(len(np.loadtxt("8x8.dat")[:,3])):
    arr = np.array([np.loadtxt("8x8.dat")[:,3][i], np.loadtxt("8x8_2.dat")[:,3][i], np.loadtxt("8x8_3.dat")[:,3][i]])/64
    M_spin_error.append(np.std(arr)/np.sqrt(3))
    

In [147]:

plt.errorbar(T, E_spin, yerr=E_spin_error, fmt='o', linestyle='-')
plt.xlabel('T/K')
plt.ylabel('Energy per spin')
plt.legend(fontsize=15)

No handles with labels found to put in legend.


In [69]:
# error bars slightly larger at phase transiton

In [78]:
plt.errorbar(T, M_spin, yerr=M_spin_error, fmt='o', linestyle='-')
plt.xlabel('T/K')
plt.ylabel('Magnetisation per spin')
plt.legend(fontsize=15)

No handles with labels found to put in legend.


In [669]:
fig, axs = plt.subplots(2)
axs[0].errorbar(T, E_spin, yerr=E_spin_error, fmt='.', linestyle='-')
axs[1].errorbar(T, M_spin, yerr=M_spin_error, fmt='.', linestyle='-')
axs[1].set_ylim(-1.5,1.5)
axs[1].set_xlabel('Temp/K')
axs[0].set_ylabel('Energy per spin')
axs[1].set_ylabel('Magnetisation per spin')

Text(0, 0.5, 'Magnetisation per spin')

In [76]:
# large errors around Tc
# magnetisation has larger errors as it fluctuates more 
# many microstates with differtent M for the same E

In [19]:
# below curie T all spins aligned therefore magnetisation per spin  = 1 
# (total spin/N)
# above Tc the magnetic spins are randomly aligned therefore their sum = 0
# magnetisation oscillates around zero
# phase transition observed

In [31]:
# magnetism rapidly drops
# this is what the curie curve is meant to look like

### Task 6

TASK 6: Repeat the final task of the previous section for the following lattice sizes: 2x2, 4x4, 8x8, 16x16, 32x32. Make sure that you name each datafile that your produce after the corresponding lattice size! Write a Python script to make a plot showing the energy per spin versus temperature for each of your lattice sizes. Hint: the NumPy loadtxt function is the reverse of the savetxt function, and can be used to read your previously saved files into the script. Repeat this for the magnetisation. As before, use the plot controls to save your a PNG image of your plot and attach this to the report. How big a lattice do you think is big enough to capture the long range correlations?

In [ ]:
# doing 3 repeats for each with 100k cycles for 8x8

In [29]:
# increasing size to correctly model long range fluctuations

In [20]:
# 2x2 4x4

<div style="display: flex">
<img src="task%206%202x2.png" width="300"/>
<img src="task%206%204x4.png" width="300"/>
</div>

In [785]:
T = np.loadtxt("2x2.dat")[:,0]
E_spin2 = np.loadtxt("2x2.dat")[:,1]/2**2
M_spin2 =  np.loadtxt("2x2.dat")[:,3]/2**2
E_spin4 = np.loadtxt("4x4.dat")[:,1]/4**2
M_spin4 =  np.loadtxt("4x4.dat")[:,3]/4**2
E_spin8 = np.loadtxt("8x8.dat")[:,1]/8**2
M_spin8 =  np.loadtxt("8x8.dat")[:,3]/8**2
E_spin16 = np.loadtxt("16x16.dat")[:,1]/16**2
M_spin16 =  np.loadtxt("16x16.dat")[:,3]/16**2
E_spin32 = np.loadtxt("32x32.dat")[:,1]/32**2
M_spin32 =  np.loadtxt("32x32.dat")[:,3]/32**2


In [225]:
fig, axs = plt.subplots(2)
axs[0].plot(T, E_spin2, label ='2x2')
axs[0].plot(T, E_spin4, label ='4x4')
axs[0].plot(T, E_spin8, label ='8x8')
axs[0].plot(T, E_spin16, label ='16x16')
axs[0].plot(T, E_spin32, label ='32x32')
axs[0].set_ylim(-2,2)
axs[0].set_xlabel('Temp/K')
axs[0].set_ylabel('Energy per spin')
axs[0].legend()

axs[1].plot(T, M_spin2, label ='2x2')
axs[1].plot(T, M_spin4, label ='4x4')
axs[1].plot(T, M_spin8, label ='8x8')
axs[1].plot(T, M_spin16, label ='16x16')
axs[1].plot(T, M_spin32, label ='32x32')
axs[1].set_ylim(-1.5,1.5)
axs[1].set_xlabel('Temp/K')
axs[1].set_ylabel('Magnetisation per spin')
#axs[1].legend(fontsize = 5)
# legend applies to both graphs

Text(0, 0.5, 'Magnetisation per spin')

In [99]:
# legend applies to both graphs
# finite size effect - 
# converges to a value 
# discontinuity at the curie temp
# at phase transition correlation length is inifnite
#

In [782]:
np.loadtxt(f"2x2.dat")[:,1]

array([-8.00034223, -8.00034223, -8.00034223, -7.99933017, -7.99680002,
       -7.98212513, -7.9699804 , -7.90116025, -7.92696781, -7.82879789,
       -7.74074858, -7.70886866, -7.64561485, -7.41891318, -7.36527395,
       -7.14464465, -7.06570389, -6.89264146, -6.76512178, -6.56827591,
       -6.52172111, -6.12246304, -5.9989916 , -6.02530518, -5.65489086,
       -5.63464964, -5.51472041, -5.19136692, -5.0517025 , -5.04056983,
       -4.83360335, -4.84169984, -4.6402997 , -4.57299764, -4.46369505,
       -4.3847543 , -4.11554607, -4.01838821, -3.95564043, -4.04773798,
       -3.77448151, -3.77245739, -3.69908297, -3.54929794, -3.46630894])

In [786]:
E_avg_list = np.zeros((5,45))
M_avg_list = np.zeros((5,45))

E_error = np.zeros((5,45))
M_error = np.zeros((5,45))
    
for i in [2,4,8,16,32]:
    file1 = np.loadtxt(f"{i}x{i}.dat")
    file2 = np.loadtxt(f"{i}x{i}_2.dat")
    file3 = np.loadtxt(f"{i}x{i}_3.dat")
    
    #var_name_Eerror = f"E_spin_error{i}"
    #locals()[var_name_Eerror] = []
    error_E = np.zeros((45))
    for j in range(len(file1[:,1])):
        arr = np.array([file1[:,1][j], file2[:,1][j], file3[:,1][j]])/i**2
        error_E[j] = np.std(arr)/np.sqrt(3)
        #locals()[var_name_Eerror].append(np.std(arr)/np.sqrt(3))
    
    #var_name_Merror = f"M_spin_error{i}"
    #locals()[var_name_Merror] = []
    error_M = np.zeros((45))
    for j in range(len(file1[:,1])):
        arr = np.array([file1[:,3][j], file2[:,3][j], file3[:,3][j]])/i**2
        error_M[j] = np.std(arr)/np.sqrt(3)
        #locals()[var_name_Merror].append(np.std(arr)/np.sqrt(3))
        
    if i == 2:
        E_avg_list[0] = (file1[:,1] + file2[:,1] + file3[:,1])/(3*2**2)
        M_avg_list[0] = (file1[:,3] + file2[:,3] + file3[:,3])/(3*2**2)
        E_error[0] = error_E
        M_error[0] = error_M
    if i == 4:
        E_avg_list[1] = (file1[:,1] + file2[:,1] + file3[:,1])/(3*4**2)
        M_avg_list[1] = (file1[:,3] + file2[:,3] + file3[:,3])/(3*4**2)
        E_error[1] = error_E
        M_error[1] = error_M
    if i == 8:
        E_avg_list[2] = (file1[:,1] + file2[:,1] + file3[:,1])/(3*64)
        M_avg_list[2] = (file1[:,3] + file2[:,3] + file3[:,3])/(3*64)
        E_error[2] = error_E
        M_error[2] = error_M
    if i == 16:
        E_avg_list[3] = (file1[:,1] + file2[:,1] + file3[:,1])/(3*16**2)
        M_avg_list[3] = (file1[:,3] + file2[:,3] + file3[:,3])/(3*16**2)
        E_error[3] = error_E
        M_error[3] = error_M
    if i == 32:
        E_avg_list[4] = (file1[:,1] + file2[:,1] + file3[:,1])/(3*32**2)
        M_avg_list[4] = (file1[:,3] + file2[:,3] + file3[:,3])/(3*32**2)
        E_error[4] = error_E
        M_error[4] = error_M
        

In [787]:
#fig, axs = plt.subplots(2)
for i in range(len(E_avg_list)):
    plt.plot(T, E_avg_list[i], label =f'{2**(i+1)}x{2**(i+1)}')
    plt.errorbar(T, E_avg_list[i], yerr=E_error[i], fmt='o', linestyle='-')

    

In [788]:
for i in range(len(E_avg_list)):
    plt.plot(T, M_avg_list[i], label =f'{2**(i+1)}x{2**(i+1)}')
    plt.errorbar(T, M_avg_list[i], yerr=M_error[i], fmt='o', linestyle='-')

In [789]:
fig, axs = plt.subplots(2)
for i in range(len(E_avg_list)):
    #axs[0].plot(T, E_avg_list[i], label =f'{2**(i+1)}x{2**(i+1)}')
    axs[0].errorbar(T, E_avg_list[i], yerr=E_error[i], fmt='.', linestyle='-', label =f'{2**(i+1)}x{2**(i+1)}')
    #axs[1].plot(T, M_avg_list[i], label =f'{2**(i+1)}x{2**(i+1)}')
    axs[1].errorbar(T, M_avg_list[i], yerr=M_error[i], fmt='.', linestyle='-')
    
axs[0].set_ylim(-2.5,0)
axs[1].set_ylim(-1.2,1.2)
axs[1].set_xlabel('Temp/K')
axs[0].set_ylabel('Energy per spin')
axs[1].set_ylabel('Magnetisation per spin')
axs[0].legend()

<div>
<img src="task%206%20with%20errors.png" width="400"/>
</div>

In [ ]:
plt.errorbar(T, E_avg_list, yerr=E_error, fmt='o', linestyle='-')
plt.xlabel('T/K')
plt.ylabel('Energy per spin')
plt.legend(fontsize=15)

In [21]:
# 16x16 32x32 (0.5-5 0.25)

<div style="display: flex">
<img src="task%206%2016x16.png" width="300"/>
<img src="task%206%2032x32%200.1.png" width="300"/>
</div>

In [30]:
# from can determine when the grid size is large enough whent the changes 
# between graphs is not significant
# looking at the graphs it looks like 16x16 is large enough to correctly model
# the long range fluctuations

In [27]:
T = np.loadtxt("2x2.dat")[:,0]
E = np.loadtxt("2x2.dat")[:,1]
M = np.loadtxt("2x2.dat")[:,3]


### Task 7

the heat capacity of the 2D Ising model should become very strongly peaked at the phase transition temperature (in fact, when T = T_C exactly, C diverges).

$$ C = \frac{\partial{\langle E \rangle}}{\partial{T}} $$
$$ C = \frac{Var[E]}{k_BT^2}

Using this relation, and the data that you generated in the previous sections, you can now determine the heat capacity of your lattice as a function of temperature and system size.

TASK 7b: Write a Python script to make a plot showing the heat capacity versus temperature for each of your lattice sizes from the previous section. You may need to do some research to recall the connection between the variance of a variable, $\mathrm{Var}[X]$, the mean of its square $\left\langle X^2\right\rangle$, and its squared mean $\left\langle X\right\rangle^2$. You may find that the data around the peak is very noisy — this is normal, and is a result of being in the critical region. As before, use the plot controls to save your a PNG image of your plot and attach this to the report.

In [290]:
T = np.loadtxt("2x2.dat")[:,0]
varE_2 = np.loadtxt("2x2.dat")[:,2]-np.loadtxt("2x2.dat")[:,1]**2
varE_4 = np.loadtxt("4x4.dat")[:,2]-np.loadtxt("4x4.dat")[:,1]**2
varE_8 = np.loadtxt("8x8.dat")[:,2]-np.loadtxt("8x8.dat")[:,1]**2
varE_16 = np.loadtxt("16x16.dat")[:,2]-np.loadtxt("16x16.dat")[:,1]**2
varE_32 = np.loadtxt("32x32.dat")[:,2]-np.loadtxt("32x32.dat")[:,1]**2

In [296]:
C_2 = varE_2/(T**2)/4
C_4 = varE_4/(T**2)/16
C_8 = varE_8/(T**2)/64
C_16 = varE_16/(T**2)/256
C_32 = varE_32/(T**2)/32**2

In [297]:
plt.plot(T, C_2, label='2x2')
plt.plot(T, C_4, label='4x4')
plt.plot(T, C_8, label='8x8')
plt.plot(T, C_16, label='16x16')
plt.plot(T, C_32, label='32x32')
plt.xlabel('Temperature/ K')
plt.ylabel("Heat capacity, C")
plt.legend()

<div>
<img src="task%207%20.png" width="400"/>
</div>

In [289]:
# strongly peaked at Tc
# peak becomes increasingly sharp as grid size increases
# infinite system: divereages at Tc

In [ ]:
# using reduced untis where kb = 1 (unitless)
# E = [K]
# T = [K]


### Task 8a

TASK 8a: A C++ program has been used to run some much longer simulations than would be possible on the college computers in Python. Each file contains six columns: T, E, E^2, M, M^2, C (the final five quantities are per spin), and you can read them with the NumPy loadtxt function as before. For each lattice size, plot the C++ data against your data. For one lattice size, save a PNG of this comparison and add it to your report — add a legend to the graph to label which is which.

In [295]:
file_2x2_c_plus = np.loadtxt("C++_data/c2x2.dat")
file_4x4_c_plus = np.loadtxt("C++_data/c4x4.dat")
file_8x8_c_plus = np.loadtxt("C++_data/c8x8.dat")
file_16x16_c_plus = np.loadtxt("C++_data/c16x16.dat")
file_32x32_c_plus = np.loadtxt("C++_data/c32x32.dat")
file_64x64_c_plus = np.loadtxt("C++_data/c64x64.dat")

T_c_plus = file_2x2_c_plus[:,0]
E_c_plus = file_2x2_c_plus[:,1]
M_c_plus = file_2x2_c_plus[:,3]

In [337]:
# energy vs T
plt.plot(file_2x2_c_plus[:,0], file_2x2_c_plus[:,1], label = 'C++')
plt.plot(T, E_avg_list[0], label ='Python')
plt.plot(file_4x4_c_plus[:,0], file_4x4_c_plus[:,1], label = 'C++')
plt.plot(T, E_avg_list[1], label ='Python')
plt.plot(file_8x8_c_plus[:,0], file_8x8_c_plus[:,1], label = 'C++')
plt.plot(T, E_avg_list[2], label ='Python')
plt.plot(file_16x16_c_plus[:,0], file_16x16_c_plus[:,1], label = 'C++')
plt.plot(T, E_avg_list[3], label ='Python')
plt.plot(file_32x32_c_plus[:,0], file_32x32_c_plus[:,1], label = 'C++')
plt.plot(T, E_avg_list[4], label ='Python')
plt.ylim(-2.2,0)
plt.legend()
plt.xlabel('Temperature/K')
plt.ylabel('E/K')



Text(0, 0.5, 'E/K')

In [338]:
# M vs T
plt.plot(file_2x2_c_plus[:,0], file_2x2_c_plus[:,3], label = 'C++')
plt.plot(T, M_avg_list[0], label ='Python')
plt.plot(file_4x4_c_plus[:,0], file_4x4_c_plus[:,3], label = 'C++')
plt.plot(T, M_avg_list[1], label ='Python')
plt.plot(file_8x8_c_plus[:,0], file_8x8_c_plus[:,3], label = 'C++')
plt.plot(T, M_avg_list[2], label ='Python')
plt.plot(file_16x16_c_plus[:,0], file_16x16_c_plus[:,3], label = 'C++')
plt.plot(T, M_avg_list[3], label ='Python')
plt.plot(file_32x32_c_plus[:,0], file_32x32_c_plus[:,3], label = 'C++')
plt.plot(T, M_avg_list[4], label ='Python')
plt.ylim(-1.5,1.5)
plt.legend(loc=1)
plt.xlabel('Temperature/K')
plt.ylabel('Magnetism')

Text(0, 0.5, 'Magnetism')

In [346]:
#8x8
fig, axs = plt.subplots(2)
axs[0].plot(file_8x8_c_plus[:,0], file_8x8_c_plus[:,1], label = 'C++')
axs[0].plot(T, E_avg_list[2], label ='Python')

axs[1].plot(file_8x8_c_plus[:,0], file_8x8_c_plus[:,3], label = 'C++')
axs[1].plot(T, M_avg_list[2], label ='Python')
axs[0].legend()

In [347]:
#16x16
fig, axs = plt.subplots(2)
axs[0].plot(file_16x16_c_plus[:,0], file_16x16_c_plus[:,1], label = 'C++')
axs[0].plot(T, E_avg_list[3], label ='Python')
axs[1].plot(file_16x16_c_plus[:,0], file_16x16_c_plus[:,3], label = 'C++')
axs[1].plot(T, M_avg_list[3], label ='Python')
axs[0].legend()

In [700]:
#32x32
data = np.loadtxt("32x32.dat")
T = data[:,0]
C = (data[:,2] - data[:,1]**2)/(T**2*32**2)

plt.plot(file_32x32_c_plus[:,0], file_32x32_c_plus[:,-1], label = 'C++')
plt.plot(T, C, label ='Python')
#plt.errorbar(T, C, yerr=E_error[4], fmt='.', linestyle='-', label =f'{2**(i+1)}x{2**(i+1)}')

plt.xlabel('Temperature/ K')
plt.ylabel("Heat capacity, C")
plt.legend()

<div>
<img src="task%208a%2032x32.png" width="400"/>
</div>

In [ ]:
#32x32
fig, axs = plt.subplots(2)
axs[0].plot(file_32x32_c_plus[:,0], file_32x32_c_plus[:,-1], label = 'C++')
#axs[0].plot(T, E_avg_list[3], label ='Python')
axs[0].errorbar(T, E_avg_list[4], yerr=E_error[4], fmt='.', linestyle='-', label =f'{2**(i+1)}x{2**(i+1)}')

axs[1].plot(file_32x32_c_plus[:,0], file_32x32_c_plus[:,3], label = 'C++')
#axs[1].plot(T, M_avg_list[4], label ='Python')
axs[1].errorbar(T, M_avg_list[4], yerr=M_error[4], fmt='.', linestyle='-', label =f'{2**(i+1)}x{2**(i+1)}')

axs[0].set_ylabel('Energy per spin')
axs[1].set_xlabel('Temperature/ K')
axs[1].set_ylabel('Magnetisation per spin')
axs[0].legend()

In [321]:
# C++ data samples every 0.02K, much more fine

### Task 8b

In [707]:
data = np.loadtxt("16x16.dat") #assume data is now a 2D array containing two columns, T and C
T = data[:,0] #get the first column
C = (data[:,2] - data[:,1]**2)/(T**2*16**2) # get the second column

#first we fit the polynomial to the data
fit = np.polyfit(T, C, 5) # fit a third order polynomial

#now we generate interpolated values of the fitted polynomial over the range of our function
T_min = np.min(T)
T_max = np.max(T)
T_range = np.linspace(T_min, T_max, 1000) #generate 1000 evenly spaced points between T_min and T_max
fitted_C_values = np.polyval(fit, T_range) # use the fit object to generate the corresponding values of C

plt.plot(T, C, marker = '.')
plt.plot(np.linspace(T_min, T_max, 1000), fitted_C_values, label = 'Polynomial fit of order 5')
#plt.plot(np.linspace(T_min, T_max, 1000), f(np.linspace(T_min, T_max, 1000), *popt))
plt.legend()
plt.ylabel("Heat Capacity, C")
plt.xlabel("Temperature/ K")


Text(0.5, 0, 'Temperature/ K')

In [377]:
np.polyfit(T, C, 10)

array([-1.02189019e+00,  2.72057637e+01, -3.11596326e+02,  2.00783239e+03,
       -7.99617667e+03,  2.03864909e+04, -3.34272284e+04,  3.46256205e+04,
       -2.16190091e+04,  7.34148831e+03, -1.54356157e+03])

### Task 8c

In [708]:
data = np.loadtxt("32x32.dat") #assume data is now a 2D array containing two columns, T and C
T = data[:,0] #get the first column
C = (data[:,2] - data[:,1]**2)/(T**2*16**2) # get the second column
varE_2 = np.loadtxt("32x32.dat")[:,2]-np.loadtxt("32x32.dat")[:,1]**2
Tmin = 2 #for example
Tmax = 3 #for example

selection = np.logical_and(T > Tmin, T < Tmax) #choose only those rows where both conditions are true
peak_T_values = T[selection]
peak_C_values = C[selection]

peak_T_range = np.linspace(Tmin, Tmax, 1000)

fit = np.polyfit(peak_T_values, peak_C_values, 5)
fitted_C_values = np.polyval(fit, peak_T_range)

plt.plot(T, C)
plt.plot(peak_T_range, fitted_C_values, label = 'Polynomial degree = 5')
plt.ylabel("Heat Capacity, C")
plt.xlabel("Temperature/ K")
plt.legend(fontsize = 10)


In [438]:
# fit resembles shape of peak well

In [418]:
Cmax = np.max(fitted_C_values)
Tmax = peak_T_range[fitted_C_values == Cmax]
Tmax

array([2.31121121])

In [440]:
Tc_array = np.zeros(5)

### 2x2

In [445]:
data = np.loadtxt("2x2.dat")
T = data[:,0]
C = (data[:,2] - data[:,1]**2)/(T**2*16**2)
varE_2 = data[:,2]-data[:,1]**2
Tmin = 1.7
Tmax = 3.7

selection = np.logical_and(T > Tmin, T < Tmax) #choose only those rows where both conditions are true
peak_T_values = T[selection]
peak_C_values = C[selection]

peak_T_range = np.linspace(Tmin, Tmax, 1000)

fit = np.polyfit(peak_T_values, peak_C_values, 5)
fitted_C_values = np.polyval(fit, peak_T_range)

Cmax = np.max(fitted_C_values)
Tmax = peak_T_range[fitted_C_values == Cmax]
Tc_array[0] = Tmax

# plt.plot(T, C)
# plt.plot(peak_T_range, fitted_C_values, label = 'Polynomial degree = 5')


### 4x4

In [452]:
data = np.loadtxt("4x4.dat")
T = data[:,0]
C = (data[:,2] - data[:,1]**2)/(T**2*16**2)
varE_2 = data[:,2]-data[:,1]**2
Tmin = 1.7
Tmax = 3.5

selection = np.logical_and(T > Tmin, T < Tmax) #choose only those rows where both conditions are true
peak_T_values = T[selection]
peak_C_values = C[selection]

peak_T_range = np.linspace(Tmin, Tmax, 1000)

fit = np.polyfit(peak_T_values, peak_C_values, 5)
fitted_C_values = np.polyval(fit, peak_T_range)

Cmax = np.max(fitted_C_values)
Tmax = peak_T_range[fitted_C_values == Cmax]
Tc_array[1] = Tmax

plt.plot(T, C)
plt.plot(peak_T_range, fitted_C_values, label = 'Polynomial degree = 5')

### 8x8

In [453]:
data = np.loadtxt("8x8.dat")
T = data[:,0]
C = (data[:,2] - data[:,1]**2)/(T**2*16**2)
varE_2 = data[:,2]-data[:,1]**2
Tmin = 1.8
Tmax = 3.2

selection = np.logical_and(T > Tmin, T < Tmax) #choose only those rows where both conditions are true
peak_T_values = T[selection]
peak_C_values = C[selection]

peak_T_range = np.linspace(Tmin, Tmax, 1000)

fit = np.polyfit(peak_T_values, peak_C_values, 5)
fitted_C_values = np.polyval(fit, peak_T_range)

Cmax = np.max(fitted_C_values)
Tmax = peak_T_range[fitted_C_values == Cmax]
Tc_array[2] = Tmax

plt.plot(T, C)
plt.plot(peak_T_range, fitted_C_values, label = 'Polynomial degree = 5')

### 16x16

In [724]:
def curie_T(size):
    data = np.loadtxt(f"{size}x{size}.dat")
    T = data[:,0]
    C = (data[:,2] - data[:,1]**2)/(T**2*16**2)
    varE_2 = data[:,2]-data[:,1]**2
    Tmin = 2.2
    Tmax = 2.8

    selection = np.logical_and(T > Tmin, T < Tmax) #choose only those rows where both conditions are true
    peak_T_values = T[selection]
    peak_C_values = C[selection]

    peak_T_range = np.linspace(Tmin, Tmax, 1000)

    fit = np.polyfit(peak_T_values, peak_C_values, 6)
    fitted_C_values = np.polyval(fit, peak_T_range)

    Cmax = np.max(fitted_C_values)
    T_c = peak_T_range[fitted_C_values == Cmax]
    
    return T_c

In [732]:

data = np.loadtxt("16x16.dat")
T = data[:,0]
C = (data[:,2] - data[:,1]**2)/(T**2*16**2)
varE_2 = data[:,2]-data[:,1]**2
Tmin = 2.2
Tmax = 2.6

selection = np.logical_and(T > Tmin, T < Tmax) #choose only those rows where both conditions are true
peak_T_values = T[selection]
peak_C_values = C[selection]

peak_T_range = np.linspace(Tmin, Tmax, 1000)
x
fit = np.polyfit(peak_T_values, peak_C_values, 6)
fitted_C_values = np.polyval(fit, peak_T_range)

Cmax = np.max(fitted_C_values)
T_c = peak_T_range[fitted_C_values == Cmax]
    

Tc_array[3] = T_c

#plt.vlines(T_c, 0, 1.76)
plt.plot(T, C, marker = '.')
plt.plot(peak_T_range, fitted_C_values, label = 'Polynomial degree = 6')
plt.legend()
plt.ylabel("Heat Capacity, C")
plt.xlabel("Temperature/ K")


/Users/joehart/opt/anaconda3/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3343: RankWarning: Polyfit may be poorly conditioned
  exec(code_obj, self.user_global_ns, self.user_ns)


Text(0.5, 0, 'Temperature/ K')

In [733]:
T_c

array([2.31931932])

In [517]:
peak_T_values

array([2.1, 2.2, 2.3, 2.4, 2.5, 2.6, 2.7, 2.8, 2.9, 3. ])

### 32x32

In [692]:
data = np.loadtxt("32x32.dat")
T = data[:,0]
C = (data[:,2] - data[:,1]**2)/(T**2*32**2)
varE_2 = data[:,2]-data[:,1]**2
Tmin = 2.1
Tmax = 2.4

selection = np.logical_and(T > Tmin, T < Tmax) #choose only those rows where both conditions are true
peak_T_values = T[selection]
peak_C_values = C[selection]

peak_T_range = np.linspace(Tmin, Tmax, 1000)

fit = np.polyfit(peak_T_values, peak_C_values, 4)
fitted_C_values = np.polyval(fit, peak_T_range)

Cmax = np.max(fitted_C_values)
T_c = peak_T_range[fitted_C_values == Cmax]
Tc_array[4] = T_c
plt.vlines(T_c, 0, 2)
#plt.axvline(Tmin)
#plt.axvline(Tmax)
plt.ylim(0,2)
plt.plot(T, C,marker='.')
plt.plot(peak_T_range, fitted_C_values, label = 'Polynomial degree = 5')

/Users/joehart/opt/anaconda3/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3343: RankWarning: Polyfit may be poorly conditioned
  exec(code_obj, self.user_global_ns, self.user_ns)


In [535]:
Tc_array

array([2.54684685, 2.48738739, 2.36756757, 2.31931932, 2.2966967 ])

In [474]:
# theoretical exact Curie temperature for the infinite 2D Ising lattice
# Tc = 2.269
# my data tends to this value

In [544]:
# Tc_L = A/L + Tc_inf
# plot Tc_L (Tc_array) vs 1/L
# fit this curve to a y = mx + c
# Tc_inf = c

fit = np.polyfit(1/np.array([2,4,8,16,32]), Tc_array, 1)
fitted_C_values = np.polyval(fit, 1/np.array([2,4,8,16,32]))

plt.plot(1/np.array([2,4,8,16,32]), Tc_array, linestyle = '', marker = '*')
plt.plot(1/np.array([2,4,8,16,32]), fit[0]/np.array([2,4,8,16,32]) + fit[1], label = '$T_{C,\infty}$ = %.2f' %(fit[1]))
plt.legend()
plt.xlabel('1/L')
plt.ylabel('$T_{C,L}$')



Text(0, 0.5, '$T_{C,L}$')

In [559]:
fit = np.polyfit(1/np.array([4,8,16,32]), Tc_array[1:], 1, cov = True)
fitted_C_values = np.polyval(fit[0], 1/np.array([4,8,16,32]))

plt.plot(1/np.array([4,8,16,32]), Tc_array[1:], linestyle = '', marker = '*')
plt.plot(1/np.array([4,8,16,32]), fit[0][0]/np.array([4,8,16,32]) + fit[0][1], label = '$T_{C,\infty}$ = %.3f $\pm$ %.3f' %(fit[0][1], fit[1][1][1]**0.5))
plt.legend()
plt.xlabel('1/L')
plt.ylabel('$T_{C,L}$')

Text(0, 0.5, '$T_{C,L}$')

In [546]:
fit = np.polyfit(1/np.array([8,16,32]), Tc_array[2:], 1)
fitted_C_values = np.polyval(fit, 1/np.array([8,16,32]))

plt.plot(1/np.array([8,16,32]), Tc_array[2:], linestyle = '', marker = '*')
plt.plot(1/np.array([8,16,32]), fit[0]/np.array([8,16,32]) + fit[1], label = '$T_{C,\infty}$ = %.2f' %(fit[1]))
plt.legend()
plt.xlabel('1/L')
plt.ylabel('$T_{C,L}$')

Text(0, 0.5, '$T_{C,L}$')

In [673]:
fit1 = np.polyfit(1/np.array([2,4,8,16,32]), Tc_array, 1, cov = True)
fitted_C_values = np.polyval(fit1[0], 1/np.array([2,4,8,16,32]))

plt.plot(1/np.array([2,4,8,16,32]), Tc_array, linestyle = '', marker = '*')
plt.plot(1/np.array([2,4,8,16,32]), fit1[0][0]/np.array([2,4,8,16,32]) + fit1[0][1], label = '$T_{C,\infty}$ = %.3f $\pm$ %.3f K' %(fit1[0][1], fit1[1][1][1]**0.5))

fit2 = np.polyfit(1/np.array([4,8,16,32]), Tc_array[1:], 1, cov = True)
fitted_C_values = np.polyval(fit2[0], 1/np.array([4,8,16,32]))

plt.plot(1/np.array([4,8,16,32]), fit2[0][0]/np.array([4,8,16,32]) + fit2[0][1], label = '$T_{C,\infty}$ = %.3f $\pm$ %.3f K' %(fit2[0][1], fit2[1][1][1]**0.5))

plt.legend()
plt.xlabel('1/L')
plt.ylabel('$T_{C,L}$/ K')

Text(0, 0.5, '$T_{C,L}$/ K')

In [547]:
fit

array([0.75824396, 2.27257257])

### Code

old energy function (slow):
 def energy(self):
    #     "Return the total energy of the current lattice configuration."
    #     energy = 0.0
    #     # can do in an L shape only taking 2 neighbours, 
        
        
    #     for i in range(self.n_rows):
    #         for j in range(self.n_cols):
    #             sum_sisj = 0
    #             # R = (i+1)%self.n_rows
    #             # L = (i-1)%self.n_rows
    #             # UP = (j+1)%self.n_cols
    #             # DOWN = (j-1)%self.n_cols
                
    #             #neighbours = np.array([self.lattice[R, j], self.lattice[L, j], self.lattice[i, UP], self.lattice[i, DOWN]])
    #             # Compute the sum of neighbouring spins
    #             sum_sisj = np.sum(self.lattice * (R + L + UP + DOWN), axis=(0, 1))
    
    #             #sum_sisj = np.sum(self.lattice[i, j] * neighbours)
    #             energy += -0.5*self.J*sum_sisj
    #     return energy

import numpy as np

class IsingLattice:

    E = 0.0
    E2 = 0.0
    M = 0.0
    M2 = 0.0

    n_cycles = 0

    def __init__(self, n_rows, n_cols):
        self.n_rows = n_rows
        self.n_cols = n_cols
        self.lattice = np.random.choice([-1,1], size=(n_rows, n_cols))
        #self.N = n_rows*n_cols
        self.J = 1
        
        def f(x, a, b, c):
            return a + b*np.exp(c*x)
        
        self.cutoff = f(n_rows, a = -2.00899300e+03, b = 8.71557693e+02, c = 1.59759578e-01)

    # def energy(self):
    #     "Return the total energy of the current lattice configuration."
    #     energy = 0.0
    #     # can do in an L shape only taking 2 neighbours, 
        
        
    #     for i in range(self.n_rows):
    #         for j in range(self.n_cols):
    #             sum_sisj = 0
    #             # R = (i+1)%self.n_rows
    #             # L = (i-1)%self.n_rows
    #             # UP = (j+1)%self.n_cols
    #             # DOWN = (j-1)%self.n_cols
                
    #             #neighbours = np.array([self.lattice[R, j], self.lattice[L, j], self.lattice[i, UP], self.lattice[i, DOWN]])
    #             # Compute the sum of neighbouring spins
    #             sum_sisj = np.sum(self.lattice * (R + L + UP + DOWN), axis=(0, 1))
    
    #             #sum_sisj = np.sum(self.lattice[i, j] * neighbours)
    #             energy += -0.5*self.J*sum_sisj
    #     return energy
    
    def energy(self):
        """
        Return the total energy of the current lattice configuration.
        """
        # Create arrays for the neighbouring spins in the lattice
        # Just 2 directions to avoid double counting
        R = np.roll(self.lattice, shift=-1, axis=0)
        UP = np.roll(self.lattice, shift=-1, axis=1)
        
        # Compute the sum of neighbouring spins
        sum_sisj = np.sum(self.lattice * (R + UP), axis=(0, 1))
        
        # Compute the energy and return it
        energy = -self.J * sum_sisj
        return energy

    def magnetisation(self):
        "Return the total magnetisation of the current lattice configuration."
        magnetisation = np.sum(self.lattice)
        
        return magnetisation

    
        

    def montecarlostep(self, T):
        self.n_cycles += 1
        #can do this in 1-2 if statements
        #generating a random number lets you test the probability distribution for one value
        
        # complete this function so that it performs a single Monte Carlo step
        energy = self.energy()
        
        #the following two lines will select the coordinates of the random spin for you
        random_i = np.random.choice(range(0, self.n_rows))
        random_j = np.random.choice(range(0, self.n_cols))
        
        # flip the spin
        self.lattice[random_i, random_j] = -1*self.lattice[random_i, random_j]

        # calc energy of new config
        new_energy = self.energy()
        E_diff = new_energy - energy
        
        #the following line will choose a random number in the rang e[0,1) for you
        random_number = np.random.random()
        
        botlzmann = np.exp(-E_diff/(T)) #*1.38e10-23
        
        if (E_diff > 0 and random_number > botlzmann):
            # reject
            self.lattice[random_i, random_j] = -1*self.lattice[random_i, random_j]
               
        else:
            # accept the copy and return the new config and magnetisation
            energy = new_energy
        
        M = self.magnetisation()
        
        
        if self.n_cycles > self.cutoff:
            self.E += energy
            self.E2 += energy**2
            self.M += M
            self.M2 += M**2
        else:
            pass
        
        return energy, M

            



    def statistics(self):
        #complete this function so that it calculates the correct values for the averages of E, E*E (E2), M, M*M (M2), and returns them with Nsteps
        #Z = (2*np.cosh(self.J/1))**(self.N)
        N = self.n_cycles
        cutoff = self.cutoff
        if N > cutoff:
            avg_E = self.E/(N-cutoff)
            avg_E2 = self.E2/(N-cutoff)
            avg_M = self.M/(N-cutoff)
            avg_M2 = self.M2/(N-cutoff)
        
        return avg_E, avg_E2, avg_M, avg_M2, N
